### ISSUE TITLE
Uber JAR executable plugin help needed

### ISSUE THREAD

**Issue by hottelet** [spring-projects/spring-boot#4324]
This project includes the following:
1. Rest Server controller
2. Application.java
3. Apache Cassandra
4. Persistence DB interface layer

Steps to Run:
1. mvn clean install the Server with the pom.xml
2. mvn clean install the Persistence with the pom.xml
3. Start Cassandra , with the run script.

Apache Support Maven Project:
https://www.dropbox.com/s/t99yyat0bk5m7v1/rest-server-generic.tar.gz?dl=0

The problem here, is that the Spring Boot plugin to generate the executable JAR file
will not generate it from the command line using mvn clean install.

This is causing a problem when a deployment of a JAR file to Amazon EC2 instance is needed.
Can you please recommend a workaround or fix to this problem?  
If this has already been documented somewhere, forgive me, but I need to see the pom.xml 
sample with running code, to be convinced that this plugin actually works.

Please someone reach out and contact me, to let me know your solution.

thank you

-Henry Hottelet
hottelet@gmail.com

**Comment by snicoll** [comment]
>The problem here, is that the Spring Boot plugin to generate the executable JAR file
will not generate it from the command line using mvn clean install.

I don't understand. 

The steps are a bit backwards. I installed the `Persistence` module, fixed the crazy `system` scope property in the main "Server" project and ran `mvn clean install` and the fat jar was generated. 

Maybe you mean executable so that you can start you app using `./rest-server-0.1.0.jar` instead of `java -jar rest-server-0.1.0.jar`? It's [a Spring Boot 1.3 feature](http://docs.spring.io/spring-boot/docs/1.3.0.RC1/maven-plugin/repackage-mojo.html#executable) and you're using 1.2

**Comment by hottelet** [comment]
Yes, I need an executable JAR , so I can run with . java -jar rest-server-0.1.0.jar.

Do you have the updated project, and a pointer to Spring Boot 1.3 with an updated pom.xml?

Please let me know.

hottelet@gmail.com

-Henry

**Comment by snicoll** [comment]
I don't see any issue: `mvn clean install` followed by `java -jar target/rest-server-0.1.0.jar` at the root of the project you shared works for me.

That being said, sorry but this isn’t the right place to get this sort of help. [As described in the guidelines for contributing](https://github.com/spring-projects/spring-boot/blob/master/CONTRIBUTING.adoc#using-github-issues), we prefer to use GitHub issues only for bugs and enhancements. For general usage questions, please use [StackOverflow](http://stackoverflow.com). Alternatively, as I believe you have a support contract with Pivotal, you may like to contact them.

**Comment by hottelet** [comment]
I tried Pivotal Support, but they were not able to answer my questions, and were unable to provide any solution.  

Thank you for answering my questions, and pointing me to Spring Boot 1.3.  

Unfortunately, it is impossible to judge from the Stackoverflow listings, which features are released, and not released.

That is why I had to escalate the issue, because I did not know which version fixed the problem.

thanks

Henry

**Comment by dsyer** [comment]
That's because there isn't a problem. `java -jar rest-server-0.1.0.jar` would work with all versions of Spring Boot. 

I recommend that you look at the existing samples and compare them carefully with your own code. Also downloading a project from start.spring.io and looking at how that is constructed would be a good idea.

In future if you could create a small sample project on github it would be easier to share changes (you'd better trim it down a bit because 66MB of Dropbox is really not necessary here).

Note that the Stackoverflow people do a great job of documenting how to post questions (e.g. http://meta.stackoverflow.com/questions/261592/how-much-research-effort-is-expected-of-stack-overflow-users?lq=1), so there is a lot of good material there to help you.

**Comment by hottelet** [comment]
Thanks for the commentary.  

I provided a testcase that doesn't provide the results you are claiming, with version 1.2.7.RELEASE.
That testcase was based on your spring boot starter project, and is very simple to read and understand.

I tried to upgrade to Spring Boot version 1.3, but it has not been released yet.

I will search stack overflow to find the spring boot version 1.3, but it might help to know when you plan on releasing it.

If you would like, I can start a github project, and post the project, so changes could be made, it is a generic, so anyone who wants a Spring Boot project, with Cassandra can use it.

**Comment by snicoll** [comment]
> Thank you for answering my questions, and pointing me to Spring Boot 1.3.

Your current project works with no change at all. Writing again what I wrote for clarification:

> Maybe you mean executable so that you can start you app using ./rest-server-0.1.0.jar instead of java -jar rest-server-0.1.0.jar? It's a Spring Boot 1.3 feature and you're using 1.2

If you want to run your app with `java -jar rest-server-0.1.0.jar` it works **now** (instead of is key here...)

**Comment by hottelet** [comment]
The problem, is when I executed the commands like you stated, 'java -jar rest-server-0.1.0.jar, this is the exception that is thrown:

"java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:53)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NoClassDefFoundError: com/persistence/pojo/Properties
	at java.lang.Class.getDeclaredConstructors0(Native Method)
	at java.lang.Class.privateGetDeclaredConstructors(Class.java:2585)
	at java.lang.Class.getDeclaredConstructors(Class.java:1906)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.determineCandidateConstructors(AutowiredAnnotationBeanPostProcessor.java:267)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.determineConstructorsFromBeanPostProcessors(AbstractAutowireCapableBeanFactory.java:1063)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1036)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:504)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:476)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:303)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:299)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:194)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:755)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:759)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:480)
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:117)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:689)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:321)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:969)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:958)
	at server.Application.main(Application.java:59)
	... 6 more
Caused by: java.lang.ClassNotFoundException: com.persistence.pojo.Properties
	at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
	at org.springframework.boot.loader.LaunchedURLClassLoader.doLoadClass(LaunchedURLClassLoader.java:177)
	at org.springframework.boot.loader.LaunchedURLClassLoader.loadClass(LaunchedURLClassLoader.java:143)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:358)"
---------------

When spring-boot:run is executed from the command line, the exceptions do not exist.

That is the dilemma, we are seeing different sets of results, and I do not know why.  

I do not think stack overflow can help with that.

You have the same files that I have.

**Comment by hottelet** [comment]
I was hoping that someone on your team might care about the inconsistencies of results, because repeatability is important.

For the record, Verizon decided to abandon development using spring boot, due to this issue.

**Comment by wilkinsona** [comment]
We do indeed care about inconsistencies and repeatability.

Your sample project was depending on the persistence project using the system scope. It is described as follows in [Maven's documentation](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html):

> **provided**
This is much like compile, but indicates you expect the JDK or a container to provide the dependency at runtime. For example, when building a web application for the Java Enterprise Edition, you would set the dependency on the Servlet API and related Java EE APIs to scope provided because the web container provides those classes. This scope is only available on the compilation and test classpath, and is not transitive.
…
**system**
This scope is similar to provided except that you have to provide the JAR which contains it explicitly. The artifact is always available and is not looked up in a repository.
…
Dependencies with the scope system are always available and are not looked up in repository. They are usually used to tell Maven about dependencies which are provided by the JDK or the VM

Your application works when you launch it using `spring-boot:run` as the system-scoped dependency is "available on the compilation and test classpath".

When you package your application as an executable jar it is not included in that jar as, by using the system scope, you have explicitly stated that the dependency will be provided by the JDK or VM. This is exactly the same behaviour as you would get when building a war file with Maven. When you run the executable jar file (`java -jar rest-server-0.1.0.jar`), the persistence jar has not been provided by the JDK or VM and a `ClassNotFoundException` results.

I'm sorry to hear that you've decided to abandon development using Spring Boot due to this issue, particularly as the issue appears to be due to a misunderstanding of Maven's scopes rather than a bug in Spring Boot.

**Comment by hottelet** [comment]
To answer your question: Verizon badly beat me up over this, and terminated my employment effective yesterday.  

So I am looking for a new job, and wouldn't have the time to dedicate to this. 

I provided you the sample project, as a service, so you can see how I did the integration between the Datastax cassandra, and open source cassandra integration.  I think that may help you in your Spring Boot 1.3 release, when you add cassandra support, because I am hoping the integration is easy to use for that.

A lot of other companies use this technology, and the government uses this too.  Wanted to let you know.

I will look at your instructions, and feedback, to try to apply them.

**Comment by hottelet** [comment]
Andy,

David Syer, was correct, generating an executable JAR file with 1.2.7.RELEASE, executed correctly.  I took the recommendation, to use the provided scope, which worked similarly to compile scope. 

I can not explain the inconsistencies that I was seeing, but an upgrade to 1.3 release, seems the most likely recommendation.

**Comment by hottelet** [comment]
Thanks @snicoll and @wilkinsona and @dsyer for your advice.  I have since updated a smaller test case integration point with Spring Boot and Cassandra. This example works with DSE Enterprise on AWS, and Spring Boot.  This example currently runs JAVA 1.7, and Spring Boot, and has dependencies on the version of Apache Cassandra builds prior to all upgrades with Spring Boot to java 1.8. That upgrade would be future work.  

What I did to fix the problem:
- flattened architecture to have Persistence 3rd party library within the project
- used <provided> reference to Persistence 3rd party library in pom.xml , instead of systemPath
  - basically, this provided command links children projects that were otherwise considered to be 3rd party libraries to the parent master project.  
 
        <dependency>
            <groupId>com.persistence</groupId>
            <artifactId>persistence</artifactId>
            <scope>provided</scope>
            <version>1.0</version>
           
        </dependency>

- mvn clean
- mvn clean install
- Included tested version of Cassandra into build

Enjoy:
- Published project with Spring-Boot and Cassandra 
- https://github.com/hottelet/spring-boot-cassandra
